contracts/staking/DelegationManager.sol

```diff
@@ -1,230 +1,281 @@
+    //@todo :add slashing, take slashedAmount into account for totalStaked
+    stakerNFT.burn(delegatorId);
+    // @todo merge delegationManager/stakeManager capital and rewards
+    require(stakeManager.delegationTransfer(delegator.reward.add(_reward), stakerNFT.ownerOf(delegatorId)),"Amount transfer failed");
+    require(token.transfer(msg.sender, amount));
+    delete delegators[delegatorId];
+    emit Unstaked(msg.sender, delegatorId, amount, totalStaked);
   }
 
   function bond(uint256 delegatorId, uint256 validatorId) public onlyDelegator(delegatorId) {
-    StakeManager stakeManager = StakeManager(registry.getStakeManagerAddress());
+    IStakeManager stakeManager = IStakeManager(registry.getStakeManagerAddress());
     uint256 currentEpoch = stakeManager.currentEpoch(); //TODO add 1
-    Delegator storage delegator = delegators[delegatorId];
 
-    require(delegator.lastValidatorEpoch == 0 || delegator.lastValidatorEpoch.add(validatorHopLimit) <= currentEpoch, "Delegation_Limit_Reached");
-
-    address validator;
-    (,,,,,,,validator,) = stakeManager.validators(validatorId);
-    require(validator != address(0x0), "Unknown validatorId or validator doesn't expect delegations");
+    if (delegators[delegatorId].bondedTo != 0) {
+      emit ReBonding(delegatorId, delegators[delegatorId].bondedTo, validatorId);
+      _unbond(delegatorId, currentEpoch, stakeManager);
+     } else {
+      emit Bonding(delegatorId, validatorId, delegators[delegatorId].amount);
+     }
+    _bond(delegatorId, validatorId, currentEpoch, stakeManager);
+  }
 
-    // for lazy unbonding
-    if (delegator.delegationStopEpoch != 0 && delegator.delegationStopEpoch < currentEpoch ) {
-      _claimRewards(delegatorId);
-      delegator.delegationStopEpoch = 0;
-      delegator.delegationStartEpoch = 0;
-    } else {
-      require((delegator.delegationStopEpoch == 0 && // TODO: handle !0
-        delegator.delegationStartEpoch == 0) &&
-        delegators[delegatorId].bondedTo == 0,
-        "Already_Bonded");
-    }
+  function _bond(uint256 delegatorId, uint256 validatorId, uint256 epoch, IStakeManager stakeManager) private {
+    require(!validatorUnbonding[validatorId], "Validator is not accepting delegation");
+    require(stakeManager.isValidator(validatorId), "Unknown validatorId or validator doesn't expect delegations");
 
-    ValidatorContract(validator).bond(delegatorId, delegator.amount, currentEpoch);
-    delegator.delegationStartEpoch = currentEpoch;
+    // require(delegator.lastValidatorEpoch.add(validatorHopLimit) <= currentEpoch, "Delegation_Limit_Reached");
+    Delegator storage delegator = delegators[delegatorId];
     delegator.bondedTo = validatorId;
-    stakeManager.updateValidatorState(validatorId, currentEpoch, int(delegator.amount));
-    delegator.lastValidatorEpoch = currentEpoch;
-    emit Bonding(delegatorId, validatorId, validator);
+    validatorDelegation[validatorId] = validatorDelegation[validatorId].add(delegator.amount);
```

Comment:
Let's say, if delegator activation epoch is not equal to current epoch, 
we are still adding delegator amount to validator.